Sew Stack first steps

lexelby opened the pull request:
This is the start of a replacement for Params: the **Sew Stack**.

The Sew Stack will allow the user to stack up layers of stitching (Stitch Layers) for each object in their design.  Unlike with Params, there will be no limit on the number, type, and order of stitching you can have for each object.  For example, you'll be able to use multiple sets of satin underlay in any order, to create cool effects.  You could do multiple passes of meander fill, even on top of normal fill if you'd like.  The sky is the limit.

This PR is a first step toward implementing the Sew Stack.  As things stand now, the Sew Stack will be alongside Params, although there will be a checkbox to stitch only the Sew Stack for a given object (not yet implemented).  This will let us slowly implement Stitch Layer types and features in successive pull requests rather than having to get everything done in one big PR.

Ultimately I'm hoping to add an option to convert the existing Params settings into Stitch Layers to help folks transition their designs over.  That will come in a later PR.  Ultimately we should have Ink/Stitch do this conversion for them so that we can get rid of all of the old Params code, but that's for much later.

My goals with the Sew Stack:

* Flexibility in type and order of stitching
  * I'm sure our users will come up with incredibly creative ways to combine Stitch Layers to do things we never imagined!
* Replace the overcomplicated, confusing, and overcrowded Params UI.
  * I'm hoping this will make Ink/Stitch much more approachable for new users.
* Reduce our extensive use of decorators since (I think) they increase the time it Python to parse our code and initialize on startup, before our main code in `inkstitch.py` ever runs.
* Working and less confusing presets system (I think presets currently don't work consistently or maybe even at all)

lexelby commented:
Here's what I have so far.  I've laid the groundwork for the Sew Stack and Stitch Layers.  I've created a single stitch layer to start with (RunningStitchLayer).  I've started on a UI to show the stitch layers and let the user drag to reorder them and double-click to rename.  It doesn't actually do anything yet.

Next up:

* add a panel on the left below the UltimateListCtrl to hold the PropertyGrid which is where the user will edit Stitch Layer settings
* wire up the simulator
* add a switch to let the user change between previewing a single layer or the entire Sew Stack in the simulator
* save the sew stack
* handle when more than one object is selected (this is tricky -- more detail on this later)

lexelby commented:
What happens when the user selects multiple objects and runs "Sew Stack..."?  In params right now this is a total mess, and I want to do better.

If the user selects multiple objects with sew stacks that have the same kinds of layers with the same names in the same order, it probably makes sense to let them continue with editing them as a group.  Making the user edit every object's sew stack one at a time is probably going to be way too painful.

What if the existing settings in some of the layers differ between users?  It might make sense to simply overwrite the entire sew stack for all selected objects when they save.  Instead, it might be better to keep track of which settings they actually changed and only apply those settings.  I'd like to do that if we can, and possibly present the user a list of changes they made that will be applied to give them a chance to cancel and go back to editing if they like.

If the user selects mulitple objects with Sew Stacks that have _different_ layers, I think we should probably not let them.  It might be cool to figure out which subsets of objects do have similar Sew Stacks and let them select one group to edit.  We can do a call to Inkscape in the background to get thumbnails for all objects and present them in the UI.

capellancitizen commented on lib/elements/element.py:
```diff
@@ -154,6 +155,16 @@ def get_multiple_int_param(self, param, default="0"):
+            return json.loads(json_value, object_hook=DotDict)
+        except json.JSONDecodeError:
+            if default is None:
+                return {}
```
Should this return a `DotDict` instance instead by default? Maybe should also only allow `DotDict` instances for default, too?

lexelby replied:
Hmmm, interesting idea, I'll give it some thought.

capellancitizen replied:
A consumer that assumes this will be a `DotDict` might crash out if it tries to manipulate the `dict` it gets in case of a default like a `DotDict`. At least, that's what I was thinking.

lexelby commented:
Now the simulator works!  You can test this by creating a path, opening the XML editor, and setting `inkstitch:sew_stack` to this:

```
{"layers": [{"layer_id": "RunningStitch", "name": "Boring running stitch", "stitch_length": 3.78}, {"layer_id": "RunningStitch", "name": "Another running stitch layer", "stitch_length": 2},  {"layer_id": "RunningStitch", "name": "third running stitch layer", "stitch_length": 3},  {"layer_id": "RunningStitch", "name": "fourth running stitch layer", "stitch_length": 4}]}
```

lexelby commented:
@capellancitizen If you're interested, contributions to this branch would be very much appreciated.  There are lots of stitch layer types to implement like zig-zag, bean stitch, satin, fill, etc.  A dialog to choose a layer type when creating a new layer would be cool too, and of course we also need to actually save the sew stack config back to the element when applying.  I'm planning to merge this when the basic flow works, and we can fill out more stitch layers in later PRs.

capellancitizen commented:
I'd be happy to contribute, though I don't expect to have a lot of time next week or so.

One thing I was wondering about the user experience: Once the old Params are removed, will the user be expected to assign a sew stack to every element they want to have embroidered, or will visible embroiderable elements without a defined sew stack get a "default" one, akin to how node_to_elements does things?

capellancitizen commented:
(Also, because one of the goals is making a better user experience: I don't have any real experience with other software other than a couple of 1h sessions of Janome Artistic Digitizer at my local library, but I'd be interested in seeing how other software does things.)

kaalleen commented:
So nice that you are willing to contribute to this one. It is very important to improve user experience for the params. They really have become overwhelming for first time users. But don't feel pressured when it comes to time, we have as much as we need.

It definitely should have a default fallback value when a user doesn't define one. Additionally the default fallback should be customizable (as many users requested).

We all do not know any other software and in fact we try to keep it that way :D We don't want to run into accusations of steeling other ideas. So we want to make up our own minds and listen to users and their needs and our own ideas what we think works best for the plugin.

lexelby commented:
Glad to hear you're interested, and trust me when I say that whatever time you care to give to this is totally fine.  We don't rush things around here. :)

My thought was to create an extension to let the user convert the Params settings to a sew stack.  After we've got that working and tested for awhile, eventually we'll do it for them automatically using our SVG versioning system so that we can get rid of the Params code.  That's down the road.

lexelby commented:
In case anyone is thinking of doing any work today or tomorrow, it might be better to hold off.  I realized my current design doesn't work well for editing multiple objects at once and I'm reworking things and will push soon.

lexelby commented:
Okay, I did a fair bit of redesign and now this thing properly handles editing multiple objects at once.  They all have to have exactly the same set of layers, hopefully that actually makes sense.  Only the properties that the user actually changes are saved back to the SVG, and they're shown in bold.  This is starting to get pretty exciting!

capellancitizen commented:
I don't feel any pressure to commit time or anything, don't worry! I'm just saying that if I am going to help with this one, I shouldn't take on (or get "assigned") anything too core unless it's something you're willing to wait at least a week for. I'll let you know when I've had time to work on it, either by posting here or by a commit :) Speaking of, would you rather I commit to this branch directly or file PRs to merge into this one?

Re: Default sew stacks: My main thought is that for WYSIWYG reasons, things the user sees should probably be embroiderable by default, so if they draw a rectangle, by default they should get a stitch plan that embroiders that rectangle without having to go into the sew stack editor. This is more of a concern for later when the sew stack replaces params altogether, though.

As for the UX: fair, I don't want to rip anyone off, of course. I'll bring up any ideas if I have them. Not the most experienced with WXWidgets, but a UI toolkit is a UI toolkit is a UI toolkit...

lexelby commented:
Oh, that's a good point about defaults.  We should probably try to make sure that we emulate what happens now: simple running stitch or fill.

Definitely just commit directly to this branch.  Pull requests on pull requests are so confusing!

lexelby commented:
I'm starting to see the kind of new features we can get out of this way of doing things.  I've added a couple of new properties that we can support on any object: `stitch_jitter` and `stitch_path_jitter`.  The first one moves stitches a random amount in any direction up to the amount the user specifies.  The second one moves randomly left or right perpendicular to the stitch path.

I can use this and the `reverse_path` property I added to do cool things like this:

![image](https://github.com/user-attachments/assets/1adebb3f-624c-4fb1-9f7a-5ecc4e01475f)

Stitch down the path, then come back and jitter randomly to the left or right.  That's this sew stack:

```json
{"layers": [{"name": "Boring running stitch", "type_name": "Running Stitch", "stitch_length": 1, "tolerance": 0.2, "stitch_length_jitter_percent": 0, "repeats": 1, "repeat_stitches": true, "reverse_path": false, "random_seed": "path918", "stitch_jitter": 0.0, "stitch_path_jitter": 0.0, "layer_id": "RunningStitch"}, {"name": "Another running stitch layer", "type_name": "Running Stitch", "stitch_length": 1.0, "tolerance": 0.2, "stitch_length_jitter_percent": 0, "repeats": 1, "repeat_stitches": true, "reverse_path": true, "random_seed": "path918", "stitch_jitter": 0.0, "stitch_path_jitter": 0.5, "layer_id": "RunningStitch"}]}
```

lexelby commented:
I'm starting to see the kind of new features we can get out of this way of doing things.  I've added a couple of new properties that we can support on any object: `stitch_jitter` and `stitch_path_jitter`.  The first one moves stitches a random amount in any direction up to the amount the user specifies.  The second one moves randomly left or right perpendicular to the stitch path.

I can use this and the `reverse_path` property I added to do cool things like this:

![image](https://github.com/user-attachments/assets/1adebb3f-624c-4fb1-9f7a-5ecc4e01475f)

Stitch down the path, then come back and jitter randomly to the left or right.  That's this sew stack:

```json
{"layers": [{"name": "Boring running stitch", "type_name": "Running Stitch", "stitch_length": 1, "tolerance": 0.2, "stitch_length_jitter_percent": 0, "repeats": 1, "repeat_stitches": true, "reverse_path": false, "random_seed": "path918", "stitch_jitter": 0.0, "stitch_path_jitter": 0.0, "layer_id": "RunningStitch"}, {"name": "Another running stitch layer", "type_name": "Running Stitch", "stitch_length": 1.0, "tolerance": 0.2, "stitch_length_jitter_percent": 0, "repeats": 1, "repeat_stitches": true, "reverse_path": true, "random_seed": "path918", "stitch_jitter": 0.0, "stitch_path_jitter": 0.5, "layer_id": "RunningStitch"}]}
```

lexelby commented:
This is probably a good place to stop, test, review, and merge.  The basic functionality and code structure seems good.  A way of handling when the user selects multiple objects with different stitch layers in their sew stacks is still needed, but that can be in a later PR.

capellancitizen commented on lib/extensions/sew_stack_editor.py:
```diff
@@ -0,0 +1,545 @@
+        self.layer_list.ClearAll()
+
+        self.layer_list.InsertColumn(0, "", format=ulc.ULC_FORMAT_RIGHT)
+        self.layer_list.InsertColumn(1, _("Type"), format=ulc.ULC_FORMAT_CENTER)
```
When I tested, I got a bunch of errors on OnPaint:
```
Traceback (most recent call last):
  File "/home/[redacted]/.local/lib/python3.10/site-packages/wx/lib/agw/ultimatelistctrl.py", line 5284, in OnPaint
    self.DrawTextFormatted(dc, text, wx.Rect(xAligned+EXTRA_WIDTH, HEADER_OFFSET_Y, cw-EXTRA_WIDTH, h-4))
TypeError: Rect(): arguments did not match any overloaded call:
  overload 1: too many arguments
  overload 2: argument 1 has unexpected type 'float'
  overload 3: argument 1 has unexpected type 'float'
  overload 4: argument 1 has unexpected type 'float'
  overload 5: argument 1 has unexpected type 'float'
  overload 6: argument 1 has unexpected type 'float'
```
This flag was the culprit, because if this is set then in `UltimateListHeaderWindow.OnPaint` it computes `xAligned` in a way that it can be a float:
```
            elif align == ULC_FORMAT_CENTER:
                xAligned = x + wcheck + (cw - wLabel)/2
```
This is a bug in ULC, really, but it still broke rendering for me. To get it to render right I had to change it to:
```
self.layer_list.InsertColumn(1, _("Type"))
```
Guessing you didn't get this error?

lexelby replied:
Weird!  I didn't get this error.  I wonder if it's about your wxPython version?  I'm on 4.2.1.

Or maybe I just get lucky and mine's always been a float that could work as an int, and wx's argument type handler will cast a float to an int if it can?

I do want this column centered, I think...

lexelby replied:
Though at some point I want to switch to showing the type as an icon, with every stitch layer type having an associated icon.  Maybe we would leave the header blank in that case and have the column be just one icon wide.

capellancitizen replied:
Odd, I'm on wxPython 4.2.1 as well. To be precise what seems to have happened here is that on your system it renders that content as having an even width, and in my case it's odd. And because of python 3's division behaviour, an int divided by an int can be a float if it's not evenly divisible...

Imo remove the center flag for now and we can submit a fix to wxwidgets for the ULC center flag.

lexelby replied:
Works for me!  I'm not at a computer right now but feel free to proceed with that.

capellancitizen reviewed (commented):
There's sure a lot of code, but it all looks reasonable enough to me. I'd like more type annotations, but that's more personal preference than the project's style.

Trying it myself, I saw an issue with the `Timer` not liking being passed floats (fixed with a commit). One major rendering issue (which was easy to patch), see comments.

lexelby commented on lib/elements/element.py:
```diff
@@ -154,6 +155,20 @@ def get_multiple_int_param(self, param, default="0"):
         return params
 
+    def get_json_param(self, param, default=None):
+        json_value = self.get_param(param, "{}")
```
You're right!  Except get_param's default arg is required(!) so that won't work.  I guess `None` is the right thing to pass, and then it looks like we'll also need to catch TypeError.

lexelby commented:
Thanks for the timer fix!  That's the last thing I was working on, guess I missed the error.

kaalleen commented:
Is someone working on it right now?
I think it'd be nice to rebase the branch to main to fix the build issues.

lexelby commented:
I'm not, feel free to rebase.

kaalleen commented:
Rebased but the newly introduced pytest routine doesn't seem to be so happy with this branch.

lexelby commented:
Ah, I see what happened there.  There's a new EmbroideryElement being returned for every object, the SewStack.

lexelby commented:
Okay, that should fix the tests.  I decided to move the SewStack _after_ the normal FillStitch/Stroke/SatinColumn objects in the list returned by `nodes_to_elements()`.  I think it's unlikely folks will use the sew stack at the same time as the old style params.  Actually that reminds me, I need to add a checkbox for "sew stack only"!

lexelby commented:
I think it's about time to merge this, unless anyone wants more time to review?

kaalleen commented:
Sorry that I didn't really had the time yet to test it and look through the code. I'd love to though.

I do love sew stack. It's a great step forward.

I ran it right now for a short test and wrote down some questions. Not all of this has to be addressed now. But I'd love if we had some answers :D

* General questions
    * Is there a plan to handle custom defaults if we only save changed params (do we?)
    * Reset to default doesn't work
    * How do we handle lists of input (such as bean stitch patterns)
    * Any way to allow people from countries with a comma as a decimal separator to use a comma instead of a point
    * Is it possible to add icons beside of the params to explain what it is doing visually
    * type_name is translatable. I think it would it make sense to save some sort of fixed class name which is independent on translations and easy searchable?

* Repeats
    * Do we need the repeat stitches checkbox
    * If we do, we should also have a general randomize checkbox
    * Can we disable list elements based on wether it is enabled or disabled
    * Can we put the general tab enabling checkboxes on top - or even better but I doubt it's possible - have it on the label level, so it clearly indicates this is meant to enable/ disable the whole block

* Description
    * Can we make it optional or at least resizable, experienced users may appreciate the extra space
    * Description is not updating for individual parameters (I think it did before?!?) It is showing the running stitch description all the time
    * I think I would like if we did add tooltips directly to the params as well

lexelby commented:
> Sorry that I didn't really had the time yet to test it and look through the code. I'd love to though.

Oh okay! I thought everyone was done.  No rush, let's leave it open.

Lots of great questions, thank you!


> * Is there a plan to handle custom defaults if we only save changed params (do we?)

Yes, we only save changed parameters (the bold ones).  I hadn't thought about saving custom defaults, but that sounds doable!

I had an idea in mind of "saved layers".  When the user goes to add a new layer, they could choose one of the layer types, or one of their "saved layers".  For example, if they create a running stitch layer they like, they could save it and it would save all of their settings with it.

Then there would be groups of layers.  We'd have built-in ones like "fill stitch with underlay" that adds two layers, one for the underlay and one for the main fill stitch.  And again, the user could save groups of layers so they could do their own reusable custom fill stitch with underlay.

> * Reset to default doesn't work

Weird, it works for me.  I think it may not be clear though that you're supposed to select a specific setting and reset that one.  Maybe the reset/undo buttons should be grayed out until the user selects one of the rows?  Or is there something else I'm missing?

> * How do we handle lists of input (such as bean stitch patterns)

wx has a built-in property editor for this!  You can see an example by running the PropertyGrid demo.  I think you can just download [this file](https://github.com/wxWidgets/Phoenix/blob/master/demo/PropertyGrid.py) and run it.  You may have to clone that whole repo in order to run that demo though.

The ArrayList property type lets you enter the list manually with comma separators or use their list editor dialog.  We may want to subclass it and tweak it to handle commas as decimal separators, or something.  Fortunately subclassing property type editors is pretty easy.

> * Any way to allow people from countries with a comma as a decimal separator to use a comma instead of a point

Yup, I think we may be able to do that.  We already have an Ink/Stitch-specific float property class, so we could handle it in there.  It'd take some experimenting.  Maybe someone has done this before.  Or maybe WX has built-in support for localizing the decimal separator?

> * Is it possible to add icons beside of the params to explain what it is doing visually

I think we could do that, though I'm not sure if it would be next to each property name or in the second cell that has the value.  Do we want to create icons for every single property, though?  That's a lot of icons...

> * type_name is translatable. I think it would it make sense to save some sort of fixed class name which is independent on translations and easy searchable?

Well, internally there's the layer_id, which is derived from the class name, so it'd be "RunningStitch" for the RunningStitchLayer class.  That's how it figures out what class to use when loading an object's sew stack.  I'm not sure if that's what you mean though... what do you mean by easily searchable?

> * Do we need the repeat stitches checkbox
> * If we do, we should also have a general randomize checkbox

Yup!  This is a new feature I added that I don't think we have in Stroke Params.  If it's checked, it repeats the _exact_ stitches, and if not, it does a new running stitch for each pass, with new randomization.  I wonder if this checkbox should be "Repeat exact stitches" to make that more clear?

With that in mind, I don't think we need the general randomize checkbox.

> * Can we disable list elements based on wether it is enabled or disabled

I think you're asking, can we grey out a property if a related/required property is not set?  Yup, definitely, and that's something I had in mind.  In fact, we can even make it so that sub-properties are nested under a main checkbox property, and disable the sub-properties if the main checkbox is unchecked, for example.

> * Can we put the general tab enabling checkboxes on top - or even better but I doubt it's possible - have it on the label level, so it clearly indicates this is meant to enable/ disable the whole block

I think we can do all of that, there really is a lot of room to do cool stuff with the property grid :)

> * Can we make the description optional or at least resizable, experienced users may appreciate the extra space

Sure!

> * Description is not updating for individual parameters (I think it did before?!?) It is showing the running stitch description all the time

It did before!  Not sure what happened there.  I'll take a look.

> * I think I would like if we did add tooltips directly to the params as well

Sure, would you want the tooltip to show the same thing as what goes in the description box?  Or an additional short/brief tooltip per property?

lexelby commented:
Okay, now the help box is resizeable too, and if they resize it, it saves the size for next time.  They can basically get rid of it by resizing it to almost zero.

Do you think the splitter between the layer list and the editor should save its size too?

kaalleen commented:
> > Sorry that I didn't really had the time yet to test it and look through the code. I'd love to though.
> 
> Oh okay! I thought everyone was done. No rush, let's leave it open.
> 

I'm not sure when I will have time. So when you feel it is ready that's totally fine with me. I really do like the idea in principle. Just a few open questions, that's it. To be honest... I might pull back from working on Ink/Stitch for a while.
 
> I had an idea in mind of "saved layers". When the user goes to add a new layer, they could choose one of the layer types, or one of their "saved layers". For example, if they create a running stitch layer they like, they could save it and it would save all of their settings with it.

If it saves all the settings , that'd be good. When they share the file, it shouldn't change anything. So they also need to share what their defaults are within the file, just in case they didn't touch a specific object with sew stack. It still needs to start with some default layer.

> Then there would be groups of layers. We'd have built-in ones like "fill stitch with underlay" that adds two layers, one for the underlay and one for the main fill stitch. And again, the user could save groups of layers so they could do their own reusable custom fill stitch with underlay.

I do like the idea.

> Weird, it works for me. I think it may not be clear though that you're supposed to select a specific setting and reset that one. Maybe the reset/undo buttons should be grayed out until the user selects one of the rows? Or is there something else I'm missing?

Oh, maybe I am then the first one to misunderstand what it really was supposed to do. I think it should be labeled clearly :D  I didn't test it that way yet, but suppose you did.

> wx has a built-in property editor for this! You can see an example by running the PropertyGrid demo. I think you can just download [this file](https://github.com/wxWidgets/Phoenix/blob/master/demo/PropertyGrid.py) and run it. You may have to clone that whole repo in order to run that demo though.

Comma separated? Hope that doesn't collide with the decimal separator plans...

> The ArrayList property type lets you enter the list manually with comma separators or use their list editor dialog. We may want to subclass it and tweak it to handle commas as decimal separators, or something. Fortunately subclassing property type editors is pretty easy.

Nice

> I think we could do that, though I'm not sure if it would be next to each property name or in the second cell that has the value. Do we want to create icons for every single property, though? That's a lot of icons...

I believe it is worth it. Maybe not for all, but for a lot. People often understand it much faster this way - and we make sure to introduce clear options, avoiding a new mess.

> Well, internally there's the layer_id, which is derived from the class name, so it'd be "RunningStitch" for the RunningStitchLayer class. That's how it figures out what class to use when loading an object's sew stack. I'm not sure if that's what you mean though... what do you mean by easily searchable?

Like in select elements, etc.

> > * Do we need the repeat stitches checkbox
> > * If we do, we should also have a general randomize checkbox
> 
> Yup! This is a new feature I added that I don't think we have in Stroke Params. If it's checked, it repeats the _exact_ stitches, and if not, it does a new running stitch for each pass, with new randomization. I wonder if this checkbox should be "Repeat exact stitches" to make that more clear?

Hm, maybe this is a randomization option then more than a repeat option?

> Sure, would you want the tooltip to show the same thing as what goes in the description box? Or an additional short/brief tooltip per property?

Short and brief :)

kaalleen commented:
> Hm, maybe this is a randomization option then more than a repeat option?

Maybe something like `randomize repeats`?

kaalleen commented:
Ok, I think to get things done, this is ready to be merged and we can update and add more functionality later.
The only thing I really wish to discuss is, if `repeat stitches` should be renamed to `randomize repeats` and  be placed into the randomization section...

lexelby commented:
> Hm, maybe this is a randomization option then more than a repeat option?

Kind of?  It's not just about randomization though.  If you have the "repeat stitches" box unchecked, it actually treats it like multiple separate running stitch paths.  It re-runs the running stitch algorithm for each pass, so the stitches might not be the same, depending on how the running stitch algorithm works in forward and reverse.  That's why I think "repeat exact stitches" may be the best way to make this clear.

lexelby commented:
I went ahead and changed it to "repeat exact stitches", but I'm happy to continue considering switching it to your way in a later PR.  I just want to get this merged now, because I've delayed for way too long at this point. ;)

lexelby commented:
Aw darn, looks like I need to adjust for the new next_element thing.  Too sleepy now, will have to try again another day :)

lexelby commented:
Gonna go ahead and merge despite the failing builds since the problem is our current lack of a windows signing certificate.
